### PR TITLE
Add encoder and decode for HyperCash addresses

### PR DESCRIPTION
**Pull request by janus**
This PR adds  functionality that would enable encoding and decoding of  HyperCash addresses

## Issue number
166

closes https://github.com/ensdomains/address-encoder/issues/166

## Description

- hcChecksum
- hcEncode
- hcCheckEncoder
- hcDecode
- checkPrefix
- hcCheckDecoder
- BN (lib)
- blake-hash(lib)

## Reference to the specification

- https://github.com/HcashOrg/hcd/blob/master/hcutil/address.go
- https://github.com/HcashOrg/hcd/blob/master/chaincfg/params.go
- https://github.com/HcashOrg/hcd/blob/73dd2af4cc3bac9b8258f2683a5540f29a994bef/hcutil/base58/base58.go
- https://github.com/HcashOrg/hcd/blob/73dd2af4cc3bac9b8258f2683a5540f29a994bef/hcutil/base58/base58check.go

## Reference to the test address.

- https://hc-explorer.h.cash/

## List of features added/changed

- HyperCash address encoding and decoding

## How Has This Been Tested?

- I followed the testing standard already created, I picked arbitrary addresses from hcash explorer and figured out how to test it

- My code didn't affect other codes it is isolated from others

## Checklist:

- [ x] My code follows the code style of this project.
- [x ] My code implements all the required features.
- [ x] I have provided the reference link to the specification I implemented.
- [x ] I have provided enough explanation about how I provided the test address

### CODE REVIEW

**Comment by makoto** [comment]
Can you resolver conflict (also please use older version of bn.js 4 so that we don't have to install two different versions to bloat file size)

File: src/index.ts [hunk 1]
```diff
@@ -530,6 +530,159 @@ function icxAddressDecoder(data: string): Buffer {
   }
 }
 
+function hcChecksum(withoutChecksum: string): Buffer {
+  const createBlakeHash = require('blake-hash')
```
Comment (makoto) [inline src/index.ts:534]: Can you use `import` instead?

**Comment by makoto** [comment]
This PR increased the size of the file from 700k to 900k. Can you think of ways to cut out the size?

**Comment by makoto** [comment]
Hi, this one is same. Are there any way making use of existing dependencies than introducing the new ones? 

Let me know if you are still keen on working on the PR. Otherwise, I will close this PR.

**Comment by janus** [comment]
> Hi, this one is same. Are there any way making use of existing dependencies than introducing the new ones?
> 
> Let me know if you are still keen on working on the PR. Otherwise, I will close this PR.

Thanks.... Working already.

**Comment by janus** [comment]
@makoto  I have reduced the size. Please review and close.

File: src/index.ts (hunk 1 above)
Comment (makoto) [reply src/index.ts:568]: Please fix this.

**Comment by makoto** [comment]
Please resolve this https://github.com/ensdomains/address-encoder/pull/245/files#r543271506

File: src/index.ts (hunk 1 above)
Comment (janus) [reply src/index.ts]: Fixed. Thanks. Merge.

File: src/@types/bs58/index.d.ts [hunk 2]
```diff
@@ -2,3 +2,5 @@ declare module 'bs58' {
   export function encode(data: any): any;
   export function decode(data: any): any;
 }
+
+declare module 'blake-hash';
```
Comment (makoto) [inline src/@types/bs58/index.d.ts:6]: src/@types directory is structured per file and I don't think `blake-hash` is part of `bs58` lib. Please put it on its own file.

Comment (janus) [reply src/@types/bs58/index.d.ts]: Thanks. Done. Merge.

**Comment by janus** [comment]
@makoto  Resolved.

File: package.json [hunk 3]
```diff
@@ -41,7 +41,9 @@
     "typescript": "^3.8.2"
   },
   "dependencies": {
+    "@types/bn.js": "^4.11.6",
     "bech32": "^1.1.3",
+    "blake-hash": "^2.0.0",
```
Comment (makoto) [inline package.json:46]: Did you add bn and blake-hash? If so you also need to commit `package-lock.json` so that we can check the actual size increase.

Comment (janus) [reply package.json:46]: @makoto Done. Review

**Comment by makoto** [comment]
File size is 345159 ->492027 , it's too big

**Pull request closed without merging**